File: trunk/android/Comp/src/org/bosik/compensation/persistence/dao/DiaryDAO.java
package org.bosik.compensation.persistence.dao;

import java.util.Date;
import java.util.List;
import org.bosik.compensation.bo.diary.DiaryPage;
import org.bosik.compensation.persistence.exceptions.CommonDAOException;

/**
 * Источник данных для дневника
 * 
 * @author Bosik
 */
public interface DiaryDAO
{
	/**
	 * Хранит пары "дата : версия"
	 * 
	 * @author Bosik
	 */
	public static class PageVersion
	{
		public Date	date;
		public int	version;

		public PageVersion(Date date, int version)
		{
			this.date = date;
			this.version = version;
		}
	}

	/**
	 * Получает информацию о страницах, изменённых после указанного времени.
	 * 
	 * @param time
	 *            Время
	 * @return Массив пар "date:version" (см. {@link PageVersion})
	 */
	public List<PageVersion> getModList(Date time) throws CommonDAOException;

	/**
	 * Получает несколько страниц дневника из БД. Если страница не существует, она будет создана.
	 * 
	 * @param dates
	 *            Даты, для которых необходимо получить страницы
	 * @return Страницы
	 * @see #getPage(Date)
	 */
	public List<DiaryPage> getPages(List<Date> dates) throws CommonDAOException;

	/**
	 * Отправляет несколько страниц дневника в БД.
	 * 
	 * @param pages
	 *            Страницы
	 * @return Успешность отправки
	 * @see #postPage(DiaryPage)
	 */
	public void postPages(List<DiaryPage> pages) throws CommonDAOException;

	/**
	 * Получает одну страницу из БД. Для получения большего количества данных используйте
	 * {@link #getPages(List)}.
	 * 
	 * @param date
	 *            Дата
	 * @return Страница
	 */
	public DiaryPage getPage(Date date) throws CommonDAOException;

	/**
	 * Отправляет одну страницу в БД. Для отправки большего количества данных используйте
	 * {@link #postPages(List)}.
	 * 
	 * @param date
	 *            Дата
	 * @return Страница
	 */
	public void postPage(DiaryPage page) throws CommonDAOException;
}

File: trunk/win32/exporter/development/src/persistence/DiaryDAO.pas
unit DiaryDAO;

interface

uses
  SysUtils, // Exception
  DiaryRoutines, // TDate
  DiaryPage;

type
  // информация для синхронизации
  TModItem = record
    Date: TDate;
    Version: integer;
  end;

  TModList = array of TModItem;

  TDateList = array of TDate;

  // DAO
  TDiaryDAO = class
  public
    { методы }
    // TODO: make it function
    procedure GetModified(Time: TDateTime; out ModList: TModList); virtual; abstract;
    // TODO: make it function
    procedure GetVersions(const Dates: TDateList; out ModList: TModList); virtual; abstract;
    // TODO: make it function
    function GetPages(const Dates: TDateList; out Pages: TDiaryPageList): boolean; virtual; abstract;
    // TODO: make it procedure
    function PostPages(const Pages: TDiaryPageList): boolean; virtual; abstract;

    { сахар }
    function GetPage(Date: TDate): TDiaryPage;
    function PostPage(Page: TDiaryPage): boolean;
  end;

implementation

{ TDiaryDAO }

{==============================================================================}
function TDiaryDAO.GetPage(Date: TDate): TDiaryPage;
{==============================================================================}
var
  DateList: TDateList;
  PageList: TDiaryPageList;
begin
  SetLength(DateList, 1);
  DateList[0] := Date;
  GetPages(DateList, PageList);
  if (Length(PageList) = 1) then
    Result := PageList[0]
  else
    raise Exception.Create('Неверная реализация TDiaryDAO: GetPages() вернул ' + IntToStr(Length(PageList)) + ' страниц вместо одной');
end;

{==============================================================================}
function TDiaryDAO.PostPage(Page: TDiaryPage): boolean;
{==============================================================================}
var
  PageList: TDiaryPageList;
begin
  SetLength(PageList, 1);
  PageList[0] := Page;
  Result := PostPages(PageList);
end;

end.
